src/main/java/org/apache/mesos/hdfs/state/LiveState.java

```diff
@@ -1,103 +1,123 @@
+    return countOfRunningTasksWith(HDFSConstants.NAME_NODE_TASKID);
   }
 
-  public boolean notInDfsHosts(String slaveId) {
-    return !taskSlaveMap.values().contains(slaveId);
+  public Protos.TaskID getFirstNameNodeTaskId() {
+    if (getNameNodeSize() >= 1) {
+      return getNamenodeTaskIds().get(0);
+    } else {
+      return null;
+    }
+  }
+
+  public Protos.TaskID getSecondNameNodeTaskId() {
+    if (getNameNodeSize() == HDFSConstants.TOTAL_NAME_NODES) {
+      return getNamenodeTaskIds().get(1);
+    } else {
+      return null;
+    }
   }
 
-  public void addTask(Protos.TaskID taskId, String hostname, String slaveId) {
-    taskHostMap.put(taskId, hostname);
-    taskSlaveMap.put(taskId, slaveId);
-    if (taskId.getValue().contains(HDFSConstants.NAME_NODE_TASKID)) {
-      nameNodeHosts.add(hostname);
-    } else if (taskId.getValue().contains(HDFSConstants.JOURNAL_NODE_ID)) {
-      journalNodeHosts.add(hostname);
+  public Protos.SlaveID getFirstNameNodeSlaveId() {
+    if (getNameNodeSize() >= 1) {
+      return getNamenodeSlaveIds().get(0);
+    } else {
+      return null;
     }
   }
 
-  public void updateTask(Protos.TaskStatus taskStatus) {
-    if (taskStatus.getTaskId().getValue().contains(HDFSConstants.NAME_NODE_TASKID)) {
-      nameNodes.add(taskStatus.getTaskId());
-    } else if (taskStatus.getTaskId().getValue()
-        .contains(HDFSConstants.JOURNAL_NODE_ID)) {
-      journalNodes.add(taskStatus.getTaskId());
+  public Protos.SlaveID getSecondNameNodeSlaveId() {
+    if (getNameNodeSize() == HDFSConstants.TOTAL_NAME_NODES) {
+      return getNamenodeSlaveIds().get(1);
+    } else {
+      return null;
     }
   }
 
-  public void removeTask(Protos.TaskStatus taskStatus) {
-    Protos.TaskID taskId = taskStatus.getTaskId();
-    journalNodeHosts.remove(taskHostMap.get(taskId));
-    nameNodeHosts.remove(taskHostMap.get(taskId));
-    taskHostMap.remove(taskId);
-    nameNodes.remove(taskId);
-    journalNodes.remove(taskId);
-    taskSlaveMap.remove(taskId);
+  private ArrayList<Protos.TaskID> getNamenodeTaskIds() {
+    ArrayList<Protos.TaskID> namenodes = new ArrayList();
+    for (Protos.TaskStatus taskStatus : runningTasks.values()) {
+      if (taskStatus.getTaskId().getValue().contains(HDFSConstants.NAME_NODE_TASKID)) {
+        namenodes.add(taskStatus.getTaskId());
+      }
+    }
+    return namenodes;
   }
 
-  public void addStagingTask(Protos.TaskID taskId) {
-    stagingTasks.add(taskId);
+  private ArrayList<Protos.SlaveID> getNamenodeSlaveIds() {
+    ArrayList<Protos.SlaveID> namenodes = new ArrayList();
+    for (Protos.TaskStatus taskStatus : runningTasks.values()) {
+      if (taskStatus.getTaskId().getValue().contains(HDFSConstants.NAME_NODE_ID)) {
+        namenodes.add(taskStatus.getSlaveId());
+      }
+    }
+    return namenodes;
   }
 
-  public void removeStagingTask(Protos.TaskID taskId) {
-    stagingTasks.remove(taskId);
+  private int countOfRunningTasksWith(final String nodeId) {
```

Comment:
Even this will be O(tasks), when it should be trivial to count the 2 or 3 NN/JNs. I'm in favor of separate lists/sets for special nodes.